File: assignment1_rt/src/distance_node2.cpp
#include <ros/ros.h>
#include <turtlesim/Pose.h>
#include <std_msgs/Float32.h>
#include <geometry_msgs/Twist.h>
#include <cmath>
#include <limits>

// Global Variables
turtlesim::Pose turtle1_pose;
turtlesim::Pose turtle2_pose;
bool turtle1_pose_received = false;
bool turtle2_pose_received = false;

const float DISTANCE_THRESHOLD = 2.0;  // Minimum Distance Threshold
const float BOUNDARY_MIN = 1.0;        // Boundary minimum
const float BOUNDARY_MAX = 10.0;       // Boundary maximum

// A callback function that receives turtle 1's pose.
void turtle1PoseCallback(const turtlesim::Pose::ConstPtr &msg) {
    turtle1_pose = *msg;
    turtle1_pose_received = true;
}

// A callback function that receives the pose of Turtle 2.
void turtle2PoseCallback(const turtlesim::Pose::ConstPtr &msg) {
    turtle2_pose = *msg;
    turtle2_pose_received = true;
}

// Function to calculate Euclidean distance
float calculateDistance(const turtlesim::Pose &pose1, const turtlesim::Pose &pose2) {
    float dx = pose1.x - pose2.x;
    float dy = pose1.y - pose2.y;
    return std::sqrt(dx * dx + dy * dy);
}

// A function to determine whether or not the object is within the boundary
bool isNearBoundary(const turtlesim::Pose &pose) {
    return pose.x < BOUNDARY_MIN || pose.x > BOUNDARY_MAX || pose.y < BOUNDARY_MIN || pose.y > BOUNDARY_MAX;
}

// A function to stop a specific turtle
void stopTurtle(ros::Publisher &pub) {
    geometry_msgs::Twist stop_cmd;
    stop_cmd.linear.x = 0.0;
    stop_cmd.angular.z = 0.0;
    pub.publish(stop_cmd);
}

int main(int argc, char **argv) {
    ros::init(argc, argv, "distance_node2");
    ros::NodeHandle nh;

    // Publisher
    ros::Publisher distance_pub = nh.advertise<std_msgs::Float32>("/turtles_distance", 10);
    ros::Publisher stop_pub_turtle1 = nh.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 10);
    ros::Publisher stop_pub_turtle2 = nh.advertise<geometry_msgs::Twist>("/turtle2/cmd_vel", 10);

    // Subscriber
    ros::Subscriber turtle1_sub = nh.subscribe("/turtle1/pose", 10, turtle1PoseCallback);
    ros::Subscriber turtle2_sub = nh.subscribe("/turtle2/pose", 10, turtle2PoseCallback);

    ros::Rate loop_rate(10);  // 10 Hz

    while (ros::ok()) {
        ros::spinOnce();

        if (turtle1_pose_received && turtle2_pose_received) {
            // Calculate distance
            float distance = calculateDistance(turtle1_pose, turtle2_pose);
            ROS_INFO("Distance between turtles: %.2f", distance);

            // Publish Distance
            std_msgs::Float32 distance_msg;
            distance_msg.data = distance;
            distance_pub.publish(distance_msg);

            // Stop if turtles are too close
            if (distance < DISTANCE_THRESHOLD) {
                ROS_WARN("Turtles are too close! Stopping both turtles.");
                stopTurtle(stop_pub_turtle1);
                stopTurtle(stop_pub_turtle2);
            }

            // Stop if turtle is close to boundary
            if (isNearBoundary(turtle1_pose)) {
                ROS_WARN("Turtle1 is near the boundary! Stopping turtle1.");
                stopTurtle(stop_pub_turtle1);
            }

            if (isNearBoundary(turtle2_pose)) {
                ROS_WARN("Turtle2 is near the boundary! Stopping turtle2.");
                stopTurtle(stop_pub_turtle2);
            }
        }

        loop_rate.sleep();
    }

    return 0;
}

File: assignment1_rt/scripts/distance_node2.py
#!/usr/bin/env python3
"""ROS node to monitor distance between turtles and enforce boundary constraints, with dynamic threshold update."""

import rospy
from turtlesim.msg import Pose
from std_msgs.msg import Float32
from geometry_msgs.msg import Twist
from math import sqrt
from typing import Optional
from assignment1_rt.srv import (
    SetThreshold,
    SetThresholdResponse,
)  # New service type for threshold update


class DistanceMonitor:
    def __init__(self) -> None:
        rospy.init_node("distance_node2", anonymous=True)

        # Thresholds (distance_threshold is now dynamic)
        self.distance_threshold: float = 2.0  # initial threshold value
        self.boundary_min: float = 1.0
        self.boundary_max: float = 10.0

        # Turtles' positions
        self.turtle1_pose: Optional[Pose] = None
        self.turtle2_pose: Optional[Pose] = None

        # Publishers
        self.distance_pub = rospy.Publisher("/turtles_distance", Float32, queue_size=10)
        self.stop_pub_turtle1 = rospy.Publisher(
            "/turtle1/cmd_vel", Twist, queue_size=10
        )
        self.stop_pub_turtle2 = rospy.Publisher(
            "/turtle2/cmd_vel", Twist, queue_size=10
        )

        # Subscribers
        rospy.Subscriber("/turtle1/pose", Pose, self.turtle1_pose_callback)
        rospy.Subscriber("/turtle2/pose", Pose, self.turtle2_pose_callback)

        # New service server for updating threshold asynchronously
        self.threshold_service = rospy.Service(
            "/set_distance_threshold", SetThreshold, self.set_threshold_callback
        )

    def turtle1_pose_callback(self, msg: Pose) -> None:
        self.turtle1_pose = msg

    def turtle2_pose_callback(self, msg: Pose) -> None:
        self.turtle2_pose = msg

    def calculate_distance(self) -> float:
        if self.turtle1_pose and self.turtle2_pose:
            dx = self.turtle1_pose.x - self.turtle2_pose.x
            dy = self.turtle1_pose.y - self.turtle2_pose.y
            return sqrt(dx**2 + dy**2)
        return float("inf")

    def enforce_constraints(self) -> None:
        if not self.turtle1_pose or not self.turtle2_pose:
            return

        distance = self.calculate_distance()
        rospy.loginfo(f"Distance between turtles: {distance:.2f}")
        self.distance_pub.publish(Float32(data=distance))

        if distance < self.distance_threshold:
            rospy.logwarn("Turtles are too close! Stopping both turtles.")
            self.stop_turtles()

        if self.is_near_boundary(self.turtle1_pose):
            rospy.logwarn("Turtle1 is near the boundary! Stopping turtle1.")
            self.stop_turtle(self.stop_pub_turtle1)

        if self.is_near_boundary(self.turtle2_pose):
            rospy.logwarn("Turtle2 is near the boundary! Stopping turtle2.")
            self.stop_turtle(self.stop_pub_turtle2)

    def is_near_boundary(self, pose: Pose) -> bool:
        return (
            pose.x < self.boundary_min
            or pose.x > self.boundary_max
            or pose.y < self.boundary_min
            or pose.y > self.boundary_max
        )

    def stop_turtles(self) -> None:
        self.stop_turtle(self.stop_pub_turtle1)
        self.stop_turtle(self.stop_pub_turtle2)

    def stop_turtle(self, pub: rospy.Publisher) -> None:
        stop_cmd = Twist()
        stop_cmd.linear.x = 0.0
        stop_cmd.angular.z = 0.0
        pub.publish(stop_cmd)

    def set_threshold_callback(
        self, req: SetThreshold.Request
    ) -> SetThreshold.Response:
        self.distance_threshold = req.threshold
        rospy.loginfo(f"Distance threshold updated to: {req.threshold}")
        return SetThresholdResponse(
            success=True, message="Threshold updated successfully."
        )

    def run(self) -> None:
        rate = rospy.Rate(10)  # 10 Hz
        while not rospy.is_shutdown():
            self.enforce_constraints()
            rate.sleep()


if __name__ == "__main__":
    try:
        monitor = DistanceMonitor()
        monitor.run()
    except rospy.ROSInterruptException:
        rospy.loginfo("Distance monitor node terminated.")

